// redux-typescript/dist/redux-typescript.js
(function(f){if(typeof exports==="object"&&typeof module!=="undefined"){module.exports=f()}else if(typeof define==="function"&&define.amd){define([],f)}else{var g;if(typeof window!=="undefined"){g=window}else if(typeof global!=="undefined"){g=global}else if(typeof self!=="undefined"){g=self}else{g=this}g.reduxTypescript = f()}})(function(){var define,module,exports;return (function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
module.exports = require('./lib/index');

},{"./lib/index":2}],2:[function(require,module,exports){
(function (global){
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _ponyfill = require('./ponyfill');

var _ponyfill2 = _interopRequireDefault(_ponyfill);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var root = undefined; /* global window */

if (typeof global !== 'undefined') {
	root = global;
} else if (typeof window !== 'undefined') {
	root = window;
}

var result = (0, _ponyfill2.default)(root);
exports.default = result;
}).call(this,typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {})
},{"./ponyfill":3}],3:[function(require,module,exports){
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.default = symbolObservablePonyfill;
function symbolObservablePonyfill(root) {
	var result;
	var _Symbol = root.Symbol;

	if (typeof _Symbol === 'function') {
		if (_Symbol.observable) {
			result = _Symbol.observable;
		} else {
			result = _Symbol('observable');
			_Symbol.observable = result;
		}
	} else {
		result = '@@observable';
	}

	return result;
};
},{}],4:[function(require,module,exports){
/**
 * @description
 * @author yangjunbao
 * @since 2016-07-14 19:57:04
 * @version 1.0.0.0
 * @file src/applyMiddleware.ts
 * @desc src/applyMiddleware.ts
 */
'use strict';
var compose_1 = require('./utils/compose');
function applyMiddleware() {
    var middlewares = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        middlewares[_i - 0] = arguments[_i];
    }
    return function (createStore) {
        return function (reducer, preloadedState, enhancer) {
            var store = createStore(reducer, preloadedState, enhancer);
            var dispatch = store.dispatch;
            var api = {
                getState: store.getState,
                dispatch: function (action) { return dispatch(action); }
            };
            var chain = middlewares.map(function (middleware) { return middleware(api); });
            // store.dispatch = compose(...chain)(store.dispatch)
            // return store
            // FIXME: WTF, this cannot pass test, but the follow passed
            dispatch = compose_1.default.apply(void 0, chain)(store.dispatch);
            store.dispatch = dispatch;
            return store;
        };
    };
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = applyMiddleware;

},{"./utils/compose":9}],5:[function(require,module,exports){
/**
 * @description
 * @author yangjunbao
 * @since 2016-07-14 20:12:03
 * @version 1.0.0.0
 * @file src/bindActionCreators.ts
 * @desc src/bindActionCreators.ts
 */
'use strict';
function bindActionCreator(actionCreator, dispatch) {
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i - 0] = arguments[_i];
        }
        return dispatch(actionCreator.apply(void 0, args));
    };
}
exports.bindActionCreator = bindActionCreator;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = function (actionCreators, dispatch) {
    if (typeof actionCreators === 'function') {
        return bindActionCreator(actionCreators, dispatch);
    }
    return Object.keys(actionCreators).reduce(function (prev, curr) {
        if (typeof actionCreators[curr] === 'function') {
            prev[curr] = bindActionCreator(actionCreators[curr], dispatch);
        }
        return prev;
    }, {});
};

},{}],6:[function(require,module,exports){
/**
 * @description
 * @author yangjunbao
 * @since 2016-07-14 21:11:18
 * @version 1.0.0.0
 * @file src/combineReducers.ts
 * @desc src/combineReducers.ts
 */
'use strict';
function combineReducers(reducers) {
    var keys = Object.keys(reducers).filter(function (key) { return typeof reducers[key] === 'function'; });
    return function combination(state, action) {
        if (state === void 0) { state = {}; }
        var changed = false;
        var newState = {};
        keys.forEach(function (name) {
            newState[name] = reducers[name](state[name], action);
            newState[name] !== state[name] && (changed = true);
        });
        return changed ? newState : state;
    };
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = combineReducers;

},{}],7:[function(require,module,exports){
/**
 * @description
 * @author yangjunbao
 * @since 2016-07-14 20:21:58
 * @version 1.0.0.0
 * @file src/createStore.ts
 * @desc src/createStore.ts
 */
'use strict';
var symbol_observable_1 = require('symbol-observable');
exports.ActionTypes = { INIT: '@@redux/INIT' };
var Store = (function () {
    function Store(reducer, preloadedState) {
        this.isDispatching = false;
        this.currListeners = [];
        this.nextListeners = this.currListeners;
        this.reducer = reducer;
        this.state = preloadedState;
        this.dispatch = this.dispatch.bind(this);
        this.getState = this.getState.bind(this);
        this.unsubscribe = this.unsubscribe.bind(this);
        this.subscribe = this.subscribe.bind(this);
        this.replaceReducer = this.replaceReducer.bind(this);
        this.observable = this.observable.bind(this);
        this.dispatch({ type: exports.ActionTypes.INIT });
        this[symbol_observable_1.default] = this.observable;
    }
    Store.prototype.dispatch = function (action) {
        if (this.isDispatching) {
            throw new Error('Reducers may not dispatch actions.');
        }
        try {
            this.isDispatching = true;
            this.state = this.reducer(this.state, action);
        }
        finally {
            this.isDispatching = false;
        }
        this.currListeners = this.nextListeners;
        this.currListeners.forEach(function (listener) { return listener(); });
        return action;
    };
    Store.prototype.ensureCanMutateNextListeners = function () {
        this.nextListeners === this.currListeners && (this.nextListeners = this.currListeners.slice());
    };
    Store.prototype.getState = function () {
        return this.state;
    };
    Store.prototype.unsubscribe = function (listener) {
        this.ensureCanMutateNextListeners();
        var index = this.nextListeners.indexOf(listener);
        ~index && this.nextListeners.splice(index, 1);
    };
    Store.prototype.subscribe = function (listener) {
        var _this = this;
        this.ensureCanMutateNextListeners();
        this.nextListeners.push(listener);
        var removed = false;
        return function () { return removed = removed ? removed : !_this.unsubscribe(listener); };
    };
    Store.prototype.replaceReducer = function (nextReducer) {
        this.reducer = nextReducer;
        this.dispatch({ type: exports.ActionTypes.INIT });
    };
    Store.prototype.observable = function () {
        var _this = this;
        return {
            subscribe: function (observer) {
                var observeState = function () {
                    observer.next && observer.next(_this.getState());
                };
                observeState();
                var unsubscribe = _this.subscribe(observeState);
                return { unsubscribe: unsubscribe };
            }
        };
    };
    return Store;
}());
exports.Store = Store;
function createStore(reducer, stateOrEnhancer, enhancer) {
    var state;
    if (typeof stateOrEnhancer === 'function') {
        enhancer = stateOrEnhancer;
    }
    else {
        state = stateOrEnhancer;
    }
    if (enhancer) {
        return enhancer(createStore)(reducer, state);
    }
    return new Store(reducer, state);
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = createStore;

},{"symbol-observable":1}],8:[function(require,module,exports){
/**
 * @description
 * @author yangjunbao
 * @since 2016-07-14 19:52:24
 * @version 1.0.0.0
 * @file src/index.ts
 * @desc src/index.ts
 */
'use strict';
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var createStore_1 = require('./createStore');
exports.createStore = createStore_1.default;
__export(require('./createStore'));
var applyMiddleware_1 = require('./applyMiddleware');
exports.applyMiddleware = applyMiddleware_1.default;
var bindActionCreators_1 = require('./bindActionCreators');
exports.bindActionCreators = bindActionCreators_1.default;
__export(require('./bindActionCreators'));
var combineReducers_1 = require('./combineReducers');
exports.combineReducers = combineReducers_1.default;

},{"./applyMiddleware":4,"./bindActionCreators":5,"./combineReducers":6,"./createStore":7}],9:[function(require,module,exports){
/**
 * @description
 * @author yangjunbao
 * @since 2016-07-14 19:57:19
 * @version 1.0.0.0
 * @file src/compose.ts
 * @desc src/compose.ts
 */
'use strict';
function default_1() {
    var funcs = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        funcs[_i - 0] = arguments[_i];
    }
    if (funcs.length === 0) {
        return function (arg) { return arg; };
    }
    if (funcs.length === 1) {
        return funcs[0];
    }
    var last = funcs.pop();
    return function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i - 0] = arguments[_i];
        }
        return funcs.reduceRight(function (prevReturn, currFn) { return currFn(prevReturn); }, last.apply(void 0, args));
    };
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;

},{}]},{},[8])(8)
});
